### ISSUE TITLE
Enhance automated testing in our development pipeline

### ISSUE THREAD

**Issue by davidc-gds** [alphagov/govuk-frontend#3041]
## What

**Timeline:** Ongoing work, with a strong focus until March 2023, and then continuing afterwards
**Priority level:** 3 – what reduces risk of inaccessible implementations

### Potential tasks

- Research new automated accessibility testing tools and determine whether they can add benefits beyond what our jest-axe automated testing is currently achieving.
- Expand automated testing to include all example code snippets for each style, component and pattern.
- Research the feasibility of testing example code snippets with dynamic JavaScript and multiple states, such as the cookie banner and character count components.

## Why
As of November 2022, there are limitations to how the team implements jest-axe as part of our automated testing:

1. The team only tests the first example code snippet for each component and pattern – most of our components and patterns have multiple example code snippets.
2. The team only tests the static HTML version of example code snippets, as found in the HTML tab – some components and patterns include JavaScript that modifies the HTML, which tests do not capture.

## Who needs to work on this
Developers, accessibility specialist

## Who needs to review this
Developers, tech leads

## Done when

- [ ] Thing to do

**Comment by davidc-gds** [comment]
Note: this is a duplicate of an older placeholder epic. I've closed the previous issue: https://github.com/alphagov/govuk-design-system/issues/1937

**Comment by davidc-gds** [comment]
Note: this replaces an other older issue related to tooling: https://github.com/alphagov/govuk-frontend/issues/1971

**Comment by NickColley** [comment]
Just so you know one limitation of jest-axe is that it runs using JSDOM so checks like colour contrast checks do not work. Moving towards something where a real browser is orchestrated (which would also allow testing between interaction states) would improve the coverage of axe-core testing.

**Comment by colinrotherham** [comment]
@davidc-gds @NickColley Have either of you tried [`@axe-core/puppeteer`](https://www.npmjs.com/package/@axe-core/puppeteer) (from Deque Systems)?

**Comment by davidc-gds** [comment]
Hey @colinrotherham I've only heard of it Puppeteer, and I'm not sure how it compares to jest-axe (or even if comparing the two would be a valid apples-to-apples comparison).

That being said, I can think of a few reasons to warrant further investigation into Puppeteer:

- It's made by Deque, the people who make axe-core
- The public npm is newer than jest-axe (Puppeteer was released 2 years ago, while jest-axe has been in govuk-frontend for at least 3 years?)
- I think puppeteer might be related or closely tied to jest in some way? I don't know much about either, but we have a line in our Package.JSON that says 'jest-environment-puppeteer'. There's also this page about using jest with puppeteer: https://jestjs.io/docs/puppeteer

**Comment by colinrotherham** [comment]
Thanks @davidc-gds it definitely looks handy

But yeah it wouldn't be like-for-like as [`jest-axe`](https://www.npmjs.com/package/jest-axe) adds nice "reporting" for us

>Just so you know one limitation of jest-axe is that it runs using JSDOM so checks like colour contrast checks do not work. Moving towards something where a real browser is orchestrated (which would also allow testing between interaction states) would improve the coverage of axe-core testing.

@NickColley I'd imagine [`jest-axe`](https://www.npmjs.com/package/jest-axe) could use [Puppeteer's **Page.setContent()**](https://pptr.dev/api/puppeteer.page.setcontent) to render static HTML just like JSDOM? Once rendered, could pass the page object into [`@axe-core/puppeteer`](https://www.npmjs.com/package/@axe-core/puppeteer) versus passing an HTML element into [`axe-core`](https://www.npmjs.com/package/axe-core)?

**Comment by NickColley** [comment]
I think unless there's a really compelling way to opt into a real browser environment I'd not change jest-axe.

Instead my gut feeling it's better to pick a good orchestration suite e.g. something like https://testcafe.io/ or https://go.cypress.io/ (used by the GOV.UK Prototype Kit project) and run axe as part of that.

- https://www.npmjs.com/package/@testcafe-community/axe
- https://www.npmjs.com/package/cypress-axe

This way you can write easier to maintain tests that interact with components in a real browser environment and run axe between the states.

**Comment by colinrotherham** [comment]
Helped out with [`@axe-core/puppeteer`](https://www.npmjs.com/package/@axe-core/puppeteer) and shared some of our Puppeteer v19 fixes:

* https://github.com/dequelabs/axe-core-npm/pull/682

After their next release we'll be free to add real browser colour contrast testing etc 😊

**Comment by davidc-gds** [comment]
@colinrotherham has another PR in the works to cover off pretty much all 3 of the potential tasks listed in this activity.

- https://github.com/alphagov/govuk-frontend/pull/3522

If the PR is successfully merged, this may represent a successful completion of this activity, at least for its initial scope.

It also looks like it will resolve this longstanding issue related to improving automated accessibility testing:

- https://github.com/alphagov/govuk-frontend/issues/1971

**Comment by colinrotherham** [comment]
We merged it @davidc-gds 😊

Would be good to confirm if we have any missing tests such as the **Cookie Banner** UI states

**Comment by davidc-gds** [comment]
Hi @colinrotherham! Where would we be able to review which tests we currently have? I don't think I'm aware of the background info on what happened with the cookie banner UI states you've mentioned.

**Comment by colinrotherham** [comment]
I've set it up to grab all the examples from the GOV.UK Frontend Review app

For example these listing pages:

* [**Accordion** examples](https://govuk-frontend-review.herokuapp.com/components/accordion)
* [**Back link** examples](https://govuk-frontend-review.herokuapp.com/components/back-link)
* [**Breadcrumbs** examples](https://govuk-frontend-review.herokuapp.com/components/breadcrumbs)
* [**Button** examples](https://govuk-frontend-review.herokuapp.com/components/button)

Bear in mind that some components have "hidden examples" which you can navigate to:

* Hidden example: [**Button** with custom attributes](https://govuk-frontend-review.herokuapp.com/components/button/attributes/preview)
* Hidden example: [**Button (input)** with attributes](https://govuk-frontend-review.herokuapp.com/components/button/input-attributes/preview)
* Hidden example: [**Button (link)** with attributes](https://govuk-frontend-review.herokuapp.com/components/button/link-attributes/preview)
* Hidden example: [**Button (link)** without `href`](https://govuk-frontend-review.herokuapp.com/components/button/no-href/preview)

You won't find these linked in the Review app, but can see them in each component's config:
[src/govuk/**components/button/button.yaml**#L152](https://github.com/alphagov/govuk-frontend/blob/main/src/govuk/components/button/button.yaml#L152)

They're typically for edge cases used by tests

>* Research the feasibility of testing example code snippets with dynamic JavaScript and multiple states, such as the cookie banner and character count components.

I mentioned the [**Cookie Banner** (see examples)](https://govuk-frontend-review.herokuapp.com/components/cookie-banner) as it's not exactly set up how you'd find it on a service and it fit the "multiple states" point you made under **Potential tasks**

Might be good to identify components like this and add Axe tests to the Full page examples instead?

https://govuk-frontend-review.herokuapp.com/full-page-examples/cookie-banner-client-side
https://govuk-frontend-review.herokuapp.com/full-page-examples/cookie-banner-essential-cookies
https://govuk-frontend-review.herokuapp.com/full-page-examples/cookie-banner-server-side

**Comment by davidc-gds** [comment]
After chatting with @colinrotherham, it appears that the initial 3 'potential tasks' are now resolved (which is amazing!)

This issue will remain open, because there are future opportunities for further enhancement. I've listed 3 potential future enhancements in the issue description.

**Comment by mgifford** [comment]
I'm a big fan of this approach. You can see our implementation here with our own accessibility site https://accessibility.civicactions.com/posts/automated-accessibility-testing-leveraging-github-actions-and-pa11y-ci-with-axe

We've also taken some steps forward with the Drupal community too https://www.thedroptimes.com/30928/drupal-takes-step-forward-in-accessibility-with-automated-testing-integration

This is all just part of shifting accessibility left.

**Comment by davidc-gds** [comment]
We're enhancing our development processes by adding the `accessibility-alt-text-bot` to our GitHub Actions.

Here's the GitHub issue where we add it:
- https://github.com/alphagov/govuk-frontend/pull/3818

[Learn more about the GitHub enhancement](https://github.blog/2023-06-12-make-your-github-projects-more-accessible-with-accessibility-alt-text-bot/).

**Comment by cmorten** [comment]
Ended up here following a nice gov.uk design system blog post. Noticed the “future” point of “Research on automated testing focussed on screen reader outputs is completed and potential options are pursued” - bit of shameless self-promotion but would be happy to talk through a package I maintain for just that https://www.guidepup.dev, and as a starter for 10 of other options in the wild there is also a list of alternatives at https://github.com/guidepup/guidepup#similar which hopefully might be of use.
